### PR TITLE
Implement sort-merge join

### LINKED ISSUE #141: Implement Sort-Merge Join
*Note*: migrated from original JIRA: https://issues.apache.org/jira/browse/ARROW-11094

The current hash join works well when one side of the join can be loaded into memory but cannot scale beyond the available RAM.

The advantage of implementing SMJ (Sort-Merge Join) is that we can sort the left and right partitions, and write the intermediate results to disk, and then stream both sides of the join by merging these sorted partitions and we do not need to load one side into memory. At most, we need to load all batches from both sides that contain the current join key values.

In order to reduce memory pressure we will want to limit the concurrency of these sort operations.

We would still want to default to hash join when we know that the build-side can fit into memory since it is more efficient than using a sort-merge join.

[https://en.wikipedia.org/wiki/Sort-merge_join]

### PR DESCRIPTION
**Pull request by richox**
# Which issue does this PR close?

Closes #141.

 # Rationale for this change

related to #1599
# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
I hope to find time to review this more carefully tomorrow

**Comment by alamb** [comment]
cc @Dandandan  and @tustvold

File: datafusion/core/src/physical_plan/mod.rs [hunk 1]
```diff
@@ -217,8 +218,9 @@ pub trait ExecutionPlan: Debug + Send + Sync {
     fn children(&self) -> Vec<Arc<dyn ExecutionPlan>>;
 
     /// Returns a new plan where all children were replaced by new plans.
+    /// The size of `children` must be equal to the size of `ExecutionPlan::children()`.
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/mod.rs:221]:
```suggestion
    /// Returns a new plan where all children were replaced by new plans.
```

**Comment by alamb** [comment]
I plan to review this PR first thing tomorrow morning US eastern time (~ 6AM or so)

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 2]
```diff
@@ -0,0 +1,1730 @@
...
+        DataType::Int16 => append_null!(Int16Builder),
+        DataType::Int32 => append_null!(Int32Builder),
+        DataType::Int64 => append_null!(Int64Builder),
+        DataType::UInt8 => append_null!(UInt8Builder),
+        DataType::UInt16 => append_null!(UInt16Builder),
+        DataType::UInt32 => append_null!(UInt32Builder),
+        DataType::UInt64 => append_null!(UInt64Builder),
+        DataType::Float32 => append_null!(Float32Builder),
+        DataType::Float64 => append_null!(Float64Builder),
+        DataType::Utf8 => append_null!(GenericStringBuilder<i32>),
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:1073]: Perhaps this should also support `LargeStringArray` as it is supported above as well

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 3]
```diff
@@ -0,0 +1,1730 @@
...
+        DataType::Int16 => append_value!(Int16Builder, Int16Array),
+        DataType::Int32 => append_value!(Int32Builder, Int32Array),
+        DataType::Int64 => append_value!(Int64Builder, Int64Array),
+        DataType::UInt8 => append_value!(UInt8Builder, UInt8Array),
+        DataType::UInt16 => append_value!(UInt16Builder, UInt16Array),
+        DataType::UInt32 => append_value!(UInt32Builder, UInt32Array),
+        DataType::UInt64 => append_value!(UInt64Builder, UInt64Array),
+        DataType::Float32 => append_value!(Float32Builder, Float32Array),
+        DataType::Float64 => append_value!(Float64Builder, Float64Array),
+        DataType::Utf8 => {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:1109]: likewise here, `LargeUtf8`

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 4]
```diff
@@ -0,0 +1,1730 @@
...
+            "| a | b | c | a  | b | c  |",
+            "+---+---+---+----+---+----+",
+            "| 1 | 4 | 7 | 10 | 1 | 70 |",
+            "| 2 | 5 | 8 | 20 | 2 | 80 |",
+            "+---+---+---+----+---+----+",
+        ];
+        assert_batches_sorted_eq!(expected, &batches);
+        Ok(())
+    }
+}
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:1730]: As this implementation matures, adding tests with string columns as join keys would be good

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 5]
```diff
@@ -0,0 +1,1730 @@
...
+            on_buffered,
+            output_buffer,
+            output_size: 0,
+            batch_size,
+            join_type,
+            join_metrics,
+        })
+    }
+
+    /// Poll next streamed row
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:525]:
```suggestion
    /// Poll next streamed row; If there is a row available, ensures that `self.streamed_batch`
    /// is populated and returns `Some(Ok(()))`. Returns `None` when the input is exhausted
```

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 6]
```diff
@@ -0,0 +1,1730 @@
...
+                    return Poll::Ready(Some(Ok(())));
+                }
+                StreamedState::Exhausted => {
+                    return Poll::Ready(None);
+                }
+            }
+        }
+    }
+
+    /// Poll next buffered batches
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:568]:
```suggestion
    /// Poll next buffered batches; If there is a row available, ensures that `self.buffered_batches` is populated
    /// and returns `Some(Ok(()))`. Returns `None` when the input is exhaused
```

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 7]
```diff
@@ -0,0 +1,1730 @@
...
+                        self.buffered_state = BufferedState::PollingRest;
+                    }
+                }
+                BufferedState::PollingFirst => match self.buffered.poll_next_unpin(cx)? {
+                    Poll::Pending => {
+                        return Poll::Pending;
+                    }
+                    Poll::Ready(None) => {
+                        self.buffered_state = BufferedState::Exhausted;
+                        return Poll::Ready(None);
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:600]: Should this return `Poll::Read(Some(()))` if `self.buffered_data.batches` is not empty?

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 8]
```diff
@@ -0,0 +1,1730 @@
...
+    fn join_partial(&mut self) -> ArrowResult<()> {
+        // decide streamed/buffered output columns by join type
+        let output_parts =
+            self.output_buffer
+                .split_at_mut(if self.join_type != JoinType::Right {
+                    self.num_streamed_columns
+                } else {
+                    self.num_buffered_columns
+                });
+        let (streamed_output, buffered_output) = if self.join_type != JoinType::Right {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:694]: I am a little confused here because I thought `JoinType::Right` always swapped the streamed / buffered outputs: 

https://github.com/apache/arrow-datafusion/pull/2242/files#diff-e7234e2d6a85330a8c23a2a2c2fbc73a383548ff2e48f65458e4f424b07df14eR171-R176

Why does this need to swap it again?

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 9]
```diff
@@ -0,0 +1,1730 @@
...
+    right_arrays: &[ArrayRef],
+    right: usize,
+    sort_options: &[SortOptions],
+    null_equals_null: bool,
+) -> ArrowResult<Ordering> {
+    let mut res = Ordering::Equal;
+    for ((left_array, right_array), sort_options) in
+        left_arrays.iter().zip(right_arrays).zip(sort_options)
+    {
+        macro_rules! compare_value {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:872]: I think @yjshen 's RowFormat will be able to hopefully make this kind of code faster and easier to follow

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 10]
```diff
@@ -0,0 +1,1730 @@
...
+    left_arrays: &[ArrayRef],
+    left: usize,
+    right_arrays: &[ArrayRef],
+    right: usize,
+) -> ArrowResult<bool> {
+    let mut is_equal = true;
+    for (left_array, right_array) in left_arrays.iter().zip(right_arrays) {
+        macro_rules! compare_value {
+            ($T:ty) => {{
+                let left_array = left_array.as_any().downcast_ref::<$T>().unwrap();
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:949]: in the future, using the arrow https://docs.rs/arrow/11.1.0/arrow/compute/kernels/comparison/fn.eq_dyn.html kernel (and then combining the bitmasks) might be faster for most joins (which are single column) rather than this custom comparison logic.

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 11]
```diff
@@ -0,0 +1,1730 @@
...
+            ("b1", &vec![4, 5, 6]),
+            ("c2", &vec![70, 80, 90]),
+        );
+
+        let on = vec![(
+            Column::new_with_schema("b1", &left.schema())?,
+            Column::new_with_schema("b1", &right.schema())?,
+        )];
+
+        let (_, batches) = join_collect(left, right, on, JoinType::Inner).await?;
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:1267]: Something I am not sure is well covered by these tests are the various corner cases of multiple batch management -- most of the tests have only a single batch of input.

I suggest adding a `fuzz` type test for the SortMergeJoin in the spirit of https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/tests/order_spill_fuzz.rs

That runs the same (logical) inputs through the sort merge join but randomizes the split into record batches (rather than one `RecordBatch` call `RecordBatch::slice()` and divide it up into smaller parts

Also, since the merge join is coalescing ranges, I think using slightly larger RecordBatches with multiple join keys would be valuable

Comment (alamb) [reply datafusion/core/src/physical_plan/sort_merge_join.rs:1267]: maybe also double checking with hash join that they get the same answers would be good

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 12]
```diff
@@ -0,0 +1,1730 @@
...
+
+        let expected = vec![
+            "+----+----+----+----+----+----+",
```
Comment (alamb) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:1270]: double checked with postgres: 

```sql
alamb=# select * from l JOIN r ON (l.b1 = r.b1);
 a1 | b1 | c1 | a2 | b1 | c2 
----+----+----+----+----+----
  1 |  4 |  7 | 10 |  4 | 70
  2 |  5 |  8 | 20 |  5 | 80
  3 |  5 |  9 | 20 |  5 | 80
(3 rows)

```

👍

**Review by alamb — APPROVED** [review]
Thank you very much @richox -- quite an impressive "first PR" 🥇 

I also found the code a joy to read (it was well commented and well structured)

I didn't have time to review every of the various state transitions in precise detail (this is a large PR!) or all the tests, but the ones I read made sense. Also the test case coverage is very a good start 

Some follow on comments / suggestions:
1. I think this code would be better named `MergeJoin` as it appears to assume the input is already sorted rather than re-sorting. It would be good to make this clear in the comments of module. 
2. I think the coverage of the various stream corner cases (`RecordBatch` boundaries) are not well covered. I think fuzz testing could help a lot
3.  It appears to me that this implements the `SortMergeJoin` operator but does not use it in any plans (yet). I wonder you can comment / link to your thoughts on how this operator will be used?


Regarding plans: I am particularly interested in dynamically switching from a `HashJoin` to `SortMergeJoin` algorithm dynamically when memory is exhausted. I have been involved in planners that get join orders / algorithm choice "wrong" due to insufficient statistics, correlated predicates, poor cost models etc, and I think dynamic behavior is the best approach to avoid such calamities. 

So in conclusion, I think this PR could be merged and we can keep iterating on it when it is part of the code base; Nice work @richox

File: datafusion/core/src/physical_plan/sort_merge_join.rs (hunk 8 above)
Comment (richox) [reply datafusion/core/src/physical_plan/sort_merge_join.rs:694]: yes. for right-join, `streamed` exactly points to the right child and `buffered` points to the left. but the output columns are still left to right. so the references to output columns also need to be swapped here.

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 13]
```diff
@@ -0,0 +1,1730 @@
...
+impl SortMergeJoinExec {
+    /// Tries to create a new [SortMergeJoinExec].
+    /// # Error
+    /// This function errors when it is not possible to join the left and right sides on keys `on`.
+    pub fn try_new(
+        left: Arc<dyn ExecutionPlan>,
+        right: Arc<dyn ExecutionPlan>,
+        on: JoinOn,
+        join_type: JoinType,
+        sort_options: Vec<SortOptions>,
```
[the remark this answers is not in the document]
Comment (richox) [reply datafusion/core/src/physical_plan/sort_merge_join.rs:83]: i guest there may be some chance to reduce extra sorting if we support different ordering for different columns. for example:
```
select
    c, d
from (
    select
        a, b, c
    from table1
    order by
        a ASC,
        b DESC
    ) t1
join (
    select
        a, b, d
    from table2
    order by
        a ASC,
        b DESC
    ) t2
on t2.a = t1.a and t2.b = t1.b
```

in the above case, column `a` and `b` are sorted in different directions. if we support different ordering, we need no extra `SortExec` before joining.

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 14]
```diff
@@ -0,0 +1,1730 @@
...
+            JoinType::Right => (
+                self.right.clone(),
+                self.left.clone(),
+                self.on.iter().map(|on| on.1.clone()).collect(),
+                self.on.iter().map(|on| on.0.clone()).collect(),
+            ),
+        };
+
+        // execute children plans
+        let streamed = CoalescePartitionsExec::new(streamed)
```
[the remark this answers is not in the document]
Comment (yjshen) [reply datafusion/core/src/physical_plan/sort_merge_join.rs:181]: Besides, I don't think CoalescePartitionsExec would preserve sort order, making merging with two pointers impossible.

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 15]
```diff
@@ -0,0 +1,1730 @@
...
+    fn metrics(&self) -> Option<MetricsSet> {
+        Some(self.metrics.clone_inner())
+    }
+
+    fn statistics(&self) -> Statistics {
+        todo!()
+    }
+}
+
+/// Metrics for SortMergeJoinExec (Not yet implemented)
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:216]: Out of date doc?

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 16]
```diff
@@ -0,0 +1,1730 @@
...
+    pub buffered: SendableRecordBatchStream,
+    /// Current processing record batch of streamed
+    pub streamed_batch: RecordBatch,
+    /// Current processing streamed join arrays
+    pub streamed_join_arrays: Vec<ArrayRef>,
+    /// Current processing row of streamed
+    pub streamed_idx: usize,
+    /// Currrent buffered data
+    pub buffered_data: BufferedData,
+    /// (used in outer join) Is current streamed row joined at least once?
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:344]: 👍

File: datafusion/core/src/physical_plan/sort_merge_join.rs (hunk 10 above)
Comment (richox) [reply datafusion/core/src/physical_plan/sort_merge_join.rs:949]: in this joining logic, comparison is always operated between one row from `streamed` and several rows from `buffered` (mostly zero or one row in real-world data). i noticed that `eq_dyn` accepts two arrays and returns another boolean array. will it introduce extra costs to create and drop the array?

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 17]
```diff
@@ -0,0 +1,1731 @@
...
+    pub fn try_new(
+        schema: SchemaRef,
+        sort_options: Vec<SortOptions>,
+        null_equals_null: bool,
+        streamed: SendableRecordBatchStream,
+        buffered: SendableRecordBatchStream,
+        on_streamed: Vec<Column>,
+        on_buffered: Vec<Column>,
+        join_type: JoinType,
+        output_buffer: Vec<Box<dyn ArrayBuilder>>,
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:492]: Possible to reuse [MutableRecordBatch](https://github.com/apache/arrow-datafusion/blob/master/datafusion/core/src/row/reader.rs#L541) in follow-up PRs?

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 18]
```diff
@@ -0,0 +1,1731 @@
...
+
+impl Stream for SMJStream {
+    type Item = ArrowResult<RecordBatch>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        self.join_metrics.join_time.timer();
+        loop {
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:384]: Nice loop with state transition 👍

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 19]
```diff
@@ -0,0 +1,1731 @@
...
+                                    self.buffered_joined = false;
+                                    self.buffered_state = BufferedState::Init;
+                                }
+                            }
+                        }
+                        SMJState::Polling
+                    };
+                }
+                SMJState::Polling => {
+                    if ![StreamedState::Exhausted, StreamedState::Ready]
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:413]: matches! macro maybe?

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 20]
```diff
@@ -0,0 +1,1731 @@
...
+    idx: usize,
+    arrays: &mut [Box<dyn ArrayBuilder>],
+) -> ArrowResult<()> {
+    if !arrays.is_empty() {
+        return batch
+            .columns()
+            .iter()
+            .zip(batch.schema().fields())
+            .enumerate()
+            .try_for_each(|(i, (column, f))| {
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:1020]: `try_for_each` result not handled?

File: datafusion/core/src/physical_plan/sort_merge_join.rs [hunk 21]
```diff
@@ -0,0 +1,1731 @@
...
+        })
+        .collect();
+    Ok(arrays)
+}
+
+/// Append one row to part of output buffer (the array builders)
+fn append_row_to_output(
+    batch: &RecordBatch,
+    idx: usize,
+    arrays: &mut [Box<dyn ArrayBuilder>],
```
Comment (yjshen) [inline datafusion/core/src/physical_plan/sort_merge_join.rs:1012]: Again, we should generalize `MutableRecordBatch` with many common usage patterns.

**Comment by yjshen** [comment]
I've just finished my first pass of review, and the overall structure looks great to me. Nice work @richox!
I plan to re-review the polling logic carefully tomorrow morning.

**Comment by yjshen** [comment]
Cc @Dandandan you might be interested in this as well.

**Review by yjshen — APPROVED** [review]
Thanks @richox. The work is solid and looks great to me!

As @alamb pointed out before, there are several follow-ups such as SMJ rename, join folder re-org, hash-join & merge-join consolidation, etc. We could do these in later PRs.

Thanks again for this epic work!

**Comment by alamb** [comment]
I agree -- thank you so much @richox 👍
